object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -338,44 +339,135 @@ private void handleAssignment(AssignmentNode node, Set<LocalVarWithTree> newDefs
-            latestAssignmentPair,
-            typeFactory.getStoreBefore(node),
-            "variable overwritten by assignment " + node.getTree());
-        newDefs.remove(latestAssignmentPair);
+        if (setContainsLatestAssignmentPair.size() == 1) {
+          checkMustCall(
+              newDefs,
+              setContainsLatestAssignmentPair,
+              typeFactory.getStoreBefore(node),
+              "variable overwritten by assignment " + node.getTree());
+          newDefs.remove(setContainsLatestAssignmentPair);
+        } else {
+          setContainsLatestAssignmentPair.remove(latestAssignmentPair);
+        }
       }
 
       // If the rhs is an ObjectCreationNode, or a MethodInvocationNode, then it adds
       // the AssignmentNode to the newDefs.
       if ((rhs instanceof ObjectCreationNode)
           || (rhs instanceof MethodInvocationNode
               && !hasNotOwningReturnType((MethodInvocationNode) rhs))) {
-        newDefs.add(
-            new LocalVarWithTree(new LocalVariable((LocalVariableNode) lhs), node.getTree()));
+        if (typeFactory.hasMustCallChoice(rhs.getTree())) {
+          Node n = getMustCallChoiceParam(rhs, newDefs);
+          if (n != null && isVarInDefs(newDefs, (LocalVariableNode) n)) {
+            getSetContainsAssignmentTreeOfVar(newDefs, (LocalVariableNode) n)
+                .add(
+                    new LocalVarWithTree(
+                        new LocalVariable((LocalVariableNode) lhs), node.getTree()));
+          }
+        } else {
+          LocalVarWithTree lhsLocalVarWithTreeNew =
+              new LocalVarWithTree(new LocalVariable((LocalVariableNode) lhs), node.getTree());
+          Set<LocalVarWithTree> newSet = new LinkedHashSet<>();
+          newSet.add(lhsLocalVarWithTreeNew);
+          newDefs.add(newSet);
+        }
       }
 
       // Ownership Transfer
       if (rhs instanceof LocalVariableNode && isVarInDefs(newDefs, (LocalVariableNode) rhs)) {
         // If the rhs is a LocalVariableNode that exists in the newDefs (Note that if a
         // localVariableNode exists in the newDefs it means it isn't assigned to a null
         // literals), then it adds the localVariableNode to the newDefs
-        newDefs.add(
-            new LocalVarWithTree(new LocalVariable((LocalVariableNode) lhs), node.getTree()));
-        newDefs.remove(getAssignmentTreeOfVar(newDefs, (LocalVariableNode) rhs));
+        Set<LocalVarWithTree> setContainsRhs =
+            getSetContainsAssignmentTreeOfVar(newDefs, (LocalVariableNode) rhs);
+        LocalVarWithTree lhsLocalVarWithTreeNew =
+            new LocalVarWithTree(new LocalVariable((LocalVariableNode) lhs), node.getTree());
+        Set<LocalVarWithTree> newSet = new LinkedHashSet<>();
+        newSet.add(lhsLocalVarWithTreeNew);
+        newDefs.add(newSet);
+        newDefs.remove(setContainsRhs);
+      }
+    }
+  }
+
+  private Node getMustCallChoiceParam(Node node, Set<Set<LocalVarWithTree>> defs) {
```

Comment:
Need to document this method, and if it always returns `@Nullable LocalVariableNode` then change the declared type